motif/match.go

```diff
@@ -10,30 +10,94 @@ import (
 	"log"
 )
 
-
-func Match(pm PositionMatrix, record fasta.Fasta, chromName string, cutoff float64, outfile string) {
-	var refPos int
+// MatchComp computes all TFBS matches in a pairwise multiFa alignment for all motifs defined in a []PositionMatrix
+// above an input propMatch (the motif score measured as a proportion of the consensus motif score).
+// MatchComp calculates the motif score for both the ref and alt sequence, as well as the difference in motif score
+// between the two sequences, and reports these values for all identified motifs in the additional fields of output bed
+// entries.
+func MatchComp(motifs []PositionMatrix, records []fasta.Fasta, chromName string, propMatch float64, outFile string, refStart int, outputAsProportion bool) {
+	var currConsensus fasta.Fasta
+	var currConsensusScore, currRefScore, currAltScore, currMotifScoreDifference float64
+	var couldScoreRefWindow, couldScoreAltWindow, couldScoreConsensus bool
+	var currCutoff float64
+	var currAnnotation []string
+	var currOutput bed.Bed
+	var currReverseMotif PositionMatrix
 	var err error
-	out := fileio.EasyCreate(outfile)
 
+	out := fileio.EasyCreate(outFile)
 
-	for alnPos := 0; alnPos < len(record.Seq); alnPos++ {
+	for _, motif := range motifs {
+		var refPos int = refStart
+		currConsensus = ConsensusSequence(motif, false)                                    //tieBreak can be false because either way we'll have the same score.
+		currConsensusScore, couldScoreConsensus = ScoreWindow(motif, currConsensus.Seq, 0) //consensus sequence is necessarily positive strand by convention.
+		if !couldScoreConsensus {
+			log.Fatalf("Error scoring consensus sequence.")
+		}
+		currCutoff = propMatch * currConsensusScore
+		currReverseMotif = ReverseComplement(motif)
 
+		for alnPos := range records[0].Seq {
+			if dna.DefineBase(records[0].Seq[alnPos]) {
+				currRefScore, couldScoreRefWindow = ScoreWindow(motif, records[0].Seq, alnPos)
+				currAltScore, couldScoreAltWindow = ScoreWindow(motif, records[1].Seq, alnPos)
+				if couldScoreRefWindow && couldScoreAltWindow && (currRefScore > currCutoff || currAltScore > currCutoff) {
+					if outputAsProportion {
+						currMotifScoreDifference = (currRefScore / currConsensusScore) - (currAltScore / currConsensusScore)
+						currAnnotation = []string{fmt.Sprintf("%f", currRefScore/currConsensusScore), fmt.Sprintf("%f", currAltScore/currConsensusScore), fmt.Sprintf("%f", currMotifScoreDifference)}
+					} else {
+						currMotifScoreDifference = currRefScore - currAltScore
```

Comment:
What about when the motifs are aligned, but do not start at the same position.  For example, I am imagining a motif where the first two bases are not very important, but then the next four bases need to be ACGT.  I am also picturing the alignment:
ACCACGT
A-CACGT
I even though both have a strong match to this motif, it starts at alignment position 1 for the reference, but 0 for the query.
Am I understanding the code correctly?  Is there a smart way to deal with this?